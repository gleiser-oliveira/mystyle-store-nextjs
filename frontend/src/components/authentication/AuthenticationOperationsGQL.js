import gql from 'graphql-tag';
import { perPage } from '../../config';

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
        signup(email: $email, name: $name, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION($email: String!, $password: String!) {
        signin(email: $email, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY { 
        me {
            id
            email
            name
            permissions
            cart {
                id
                quantity
                product {
                    id
                    price
                    title
                    image
                    description
                }
            }
        }
    }
`;

export const REQUEST_PASSWORD_RESET_MUTATION = gql`
    mutation REQUEST_PASSWORD_RESET_MUTATION($email: String!) {
        requestPasswordReset(email: $email) {
            message
        }
    }
`;

export const RESET_PASSWORD_MUTATION = gql`
    mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
        resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
            id
            name
            email
        }
    }
`;

export const HAS_PERMISSION_QUERY = gql`
    query { 
        users {
            id
            email
            name
            permissions
        }
    }
`;

export const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY($skip: Int = 0, $first: Int = ${perPage}) { 
        users(first: $first, skip: $skip, orderBy: name_ASC) {
            id
            email
            name
            permissions
        }
    }
`;

export const USERS_PAGINATION_QUERY = gql`
    query USERS_PAGINATION_QUERY {
        usersConnection {
            aggregate {
                count
            }
        }
    }
`;

export const ALL_PERMISSIONS_QUERY = gql`
    query { 
        permissions {
          id
          permission
          description
        }
    }
`;

export const UPDATE_USER_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_USER_PERMISSIONS_MUTATION($id: ID! $permissions: [Permission]) {
        updateUserPermissions(id: $id permissions: $permissions) {
            id
            email
            name
            permissions
        }
    }
`;
