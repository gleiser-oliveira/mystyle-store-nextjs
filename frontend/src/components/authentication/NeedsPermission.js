import React from 'react';
import PropTypes from 'prop-types';

const NeedsPermission = props => {
    if (!props.user)
        return null;
    if (props.user.permissions.includes('ADMIN'))
        return props.children;
    const message = props.showMessage ? <div><p>You have no permission to access this resource.</p></div> : null;
    const fragment = ((!props.ownerId) || (props.ownerId === props.user.id)) && props.user.permissions.some(permission => props.neededPermission.includes(permission)) ? props.children : message;
    return fragment;
};

NeedsPermission.propTypes = {
    user: PropTypes.object,
    ownerId: PropTypes.string,
    neededPermission: PropTypes.array,
};

export default NeedsPermission;
