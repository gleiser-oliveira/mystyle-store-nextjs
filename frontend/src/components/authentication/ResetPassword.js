import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { Form } from '../theme/generalStyles';
import { REQUEST_PASSWORD_RESET_MUTATION, RESET_PASSWORD_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';
import { ErrorMessage, ConfirmationMessage } from '../message/Message';

class ResetPassword extends Component {
    state = {
        email: this.props.email || '',
        password: '',
        confirmPassword: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }

    handleRequestPasswordReset = async(e, requestReset) => {
        e.preventDefault();
        await requestReset();
        this.setState({
            email: ''
        });
    }

    handlePasswordReset = async(e, resetPassword) => {
        e.preventDefault();
        await resetPassword();
        this.setState({
            password: '',
            confirmPassword: '',
        });
    }

    renderRequestResetPasswordForm() {
        return (
            <Mutation mutation={REQUEST_PASSWORD_RESET_MUTATION} variables={{ email: this.state.email }} >
                {(requestReset, payload) => (
                    <Form method="post" onSubmit={e => { this.handleRequestPasswordReset(e, requestReset); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Reset your Account&rsquo;s Password</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && !payload.loading && payload.called && payload.data ? <ConfirmationMessage message={payload.data.requestPasswordReset.message} /> : null}
                            <label htmlFor="email">
                                Email
                                <input
                                    type="email"
                                    name="email"
                                    placeholder="email"
                                    value={this.state.email}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <button type="submit">Reset</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }

    renderResetPasswordForm() {
        return (
            <Mutation
                mutation={RESET_PASSWORD_MUTATION}
                variables={{
                    resetToken: this.props.token,
                    password: this.state.password,
                    confirmPassword: this.state.confirmPassword
                }}
                refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                {(resetPassword, payload) => (
                    <Form method="post" onSubmit={e => { this.handlePasswordReset(e, resetPassword); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Reset your Account&rsquo;s Password</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && payload.data ? <ConfirmationMessage message={`Password sucessfully reset for user ${payload.data.resetPassword.name}`} /> : null}
                            <label htmlFor="password">
                                Password:
                                <input
                                    type="password"
                                    name="password"
                                    placeholder="New password"
                                    value={this.state.password}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <label htmlFor="confirmPassword">
                                Confirm the new password:
                                <input
                                    type="password"
                                    name="confirmPassword"
                                    placeholder="Confirm your password"
                                    value={this.state.confirmPassword}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <button type="submit">Confirm</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }

    render() {
        return this.props.token ? this.renderResetPasswordForm() : this.renderRequestResetPasswordForm();
    }
}

ResetPassword.propTypes = {
    email: PropTypes.string,
    token: PropTypes.string,
};

export default ResetPassword;
