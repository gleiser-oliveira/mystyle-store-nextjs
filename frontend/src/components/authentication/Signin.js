import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Link from 'next/link';
import { Form } from '../theme/generalStyles';
import { SIGNIN_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';
import { ErrorMessage, ConfirmationMessage } from '../message/Message';

class Signin extends Component {
    state = {
        name: '',
        email: '',
        password: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }

    handleSubmit = async(e, signin) => {
        e.preventDefault();
        const res = await signin();
        this.setState({
            name: '',
            email: '',
            password: '',
            message: `User ${res.data.signin.name} logged successfully!`,
        });
    }

    render() {
        return (
            <Mutation mutation={SIGNIN_MUTATION} variables={{ email: this.state.email, password: this.state.password }} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                {(signin, payload) => (
                    <Form method='post' onSubmit={e => { this.handleSubmit(e, signin); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Sign into your Account</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && this.state.message ? <ConfirmationMessage message={this.state.message} /> : null}
                            <label htmlFor='email'>
                                Email
                                <input
                                    type='email'
                                    name='email'
                                    placeholder='email'
                                    value={this.state.email}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <label htmlFor='password'>
                                Password
                                <input
                                    type='password'
                                    name='password'
                                    placeholder='password'
                                    value={this.state.password}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <Link href={{
                                pathname: '/resetPassword',
                                query: { email: this.state.email }
                            }}>
                                <a>Forgot your password? Click here!</a>
                            </Link>
                            <button type='submit'>Sign In!</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }
}

export default Signin;
