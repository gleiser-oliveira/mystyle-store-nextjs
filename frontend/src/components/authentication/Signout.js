import React from 'react';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import { SIGNOUT_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';

const Signout = props => (
    <Mutation mutation={SIGNOUT_MUTATION} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
        {(signout, payload) => (
            <button onClick={async e => {
                // const res = await signout();
                await signout();
                Router.push({
                    pathname: '/products',
                });
            }}>Sign Out</button>
        )}
    </Mutation>
);

export default Signout;
