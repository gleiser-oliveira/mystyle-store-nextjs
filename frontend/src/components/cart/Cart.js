import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query, Mutation } from 'react-apollo';
import { CartStyle, Supreme, CloseButton } from './styles';
import CartProduct from './CartProduct';
import Payment from '../payment/Payment';
import { calcTotalCartPrice, formatPrice, numberOfProducts } from '../../lib/cartHelper';
import { CLEAR_CART_MUTATION } from './CartOperationsGQL';
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../../lib/ClientGQL/LocalGQL';
import { CURRENT_USER_QUERY } from '../authentication/AuthenticationOperationsGQL';

class Cart extends Component {
    clearCartAction = (e, clearCartMutation) => {
        const confirmed = confirm('Are you sure you want to empty your cart?');
        if (confirmed)
            clearCartMutation();
    }

    render() {
        return (!this.props.userCart ? null
            : (
                <Mutation mutation={TOGGLE_CART_MUTATION}>
                    {toggleCart => (
                        <Query query={LOCAL_STATE_QUERY}>
                            {payload => (
                                <CartStyle open={payload.data.cartOpen}>
                                    <header>
                                        <CloseButton onClick={toggleCart} title='close'>&times;</CloseButton>
                                        <Supreme>Your Cart</Supreme>
                                        <p>You have {numberOfProducts(this.props.userCart)} products in your cart.</p>
                                    </header>
                                    <ul>
                                        {this.props.userCart.map(cartProduct => <CartProduct key={cartProduct.id} cartProduct={cartProduct} />)}
                                    </ul>
                                    <Mutation mutation={CLEAR_CART_MUTATION} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                                        {(clearCart, payload) => (
                                            <div>
                                                <button className='clear-cart-button' disabled={payload.loading || this.props.userCart.length < 1} onClick={e => this.clearCartAction(e, clearCart)}>
                                                    <span>Clear{payload.loading && 'ing'} Cart</span>
                                                </button>
                                            </div>
                                        )}
                                    </Mutation>
                                    <footer>
                                        <div>
                                            <p>Total: {formatPrice(calcTotalCartPrice(this.props.userCart))}</p>
                                            <Payment>
                                                <button disabled={this.props.userCart.length < 1 || this.props.userCart.some(cartProduct => !cartProduct.product)}>
                                                    Checkout
                                                </button>
                                            </Payment>
                                        </div>
                                    </footer>
                                </CartStyle>
                            )}
                        </Query>
                    )}
                </Mutation>
            )
        );
    }
}

Cart.propTypes = {
    userCart: PropTypes.array,
};

export default Cart;
