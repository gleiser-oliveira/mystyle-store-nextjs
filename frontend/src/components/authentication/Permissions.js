import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NeedsPermission from './NeedsPermission';
import { Query, Mutation } from 'react-apollo';
import { ALL_PERMISSIONS_QUERY, ALL_USERS_QUERY, USERS_PAGINATION_QUERY, UPDATE_USER_PERMISSIONS_MUTATION } from './AuthenticationOperationsGQL';
import { ErrorMessage } from '../message/Message';
import Pagination from '../pagination/Pagination';
import { Center } from '../theme/generalStyles';
import { PermissionsGrid, AdminView } from './styles';
import { perPage } from '../../config';

class UserPermissions extends Component {
    state = {
        id: this.props.user.id,
        permissions: this.props.user.permissions,
    }

    handlePermissionChange = (e) => {
        let updatedPermissions = [...this.state.permissions];
        if (e.target.checked)
            updatedPermissions.push(e.target.value);
        else
            updatedPermissions = updatedPermissions.filter(permission => (permission !== e.target.value));

        this.setState({ permissions: updatedPermissions });
    }

    updatePermissionAction = (e, updatePermissions) => {
        updatePermissions();
    }

    render() {
        const permissionsInputs = this.props.allPermissions.map(permObj => (<div key={`${this.props.user.id}-permission-${permObj.permission}`} className="perm">
            <label htmlFor={`${this.props.user.id}-permission-${permObj.permission}`}>
                <input id={`${this.props.user.id}-permission-${permObj.permission}`} type="checkbox" value={permObj.permission} checked={this.state.permissions.includes(permObj.permission)} onChange={this.handlePermissionChange} />
            </label>
        </div>));

        const permissionsRow = <Mutation mutation={UPDATE_USER_PERMISSIONS_MUTATION} variables={{ id: this.state.id, permissions: this.state.permissions }}>
            {(updatePermissions, payload) => (
                [permissionsInputs, <div key={this.props.user.id} className='perm'><button onClick={(e) => { this.updatePermissionAction(e, updatePermissions); }}>UPDAT{payload.loading ? 'ING' : 'E'}</button></div>]
            )}
        </Mutation>;

        return permissionsRow;
    }
}

class Permissions extends Component {
    state = {
        pagLoading: false,
    }

    setLoading = (pag, isLoading) => {
        if (this.state[pag] !== isLoading)
            this.setState({ [pag]: isLoading });
    };

    renderPermissionsGrid = (permissionsData, usersData) => {
        return <Center>
            <AdminView>
                <PermissionsGrid>
                    <div className="col-u"></div>
                    <div className="col-p">
                        <div className="perms">
                            {permissionsData.map(perm => (<div key={perm.id} className="perm">{perm.description}</div>))}
                            <div className="perm">
                                <p>UPDATE</p>
                            </div>
                        </div>
                    </div>
                    <div className="col col-u">
                        {usersData.map(user => (<div key={user.id} className="user">{user.name}</div>))}
                    </div>
                    <div className="col col-p">
                        {usersData.map(user => (<div key={user.id} className="perms"><UserPermissions key={user.id} allPermissions={permissionsData} user={user} /></div>))}
                    </div>
                </PermissionsGrid>
            </AdminView>
            <Pagination
                page={this.props.page}
                loading={{ event: this.setLoading, name: 'pagLoading' }}
                query={USERS_PAGINATION_QUERY}
                nameObjects='Users'
                pathname='permissions'
                connectionName='usersConnection'
            />
        </Center>;
    }

    render() {
        return (
            <NeedsPermission neededPermission={['ADMIN']} user={this.props.userData} showMessage={true} >
                <Query query={ALL_PERMISSIONS_QUERY}>
                    {({ data, loading, error }) => {
                        if (loading)
                            return <p>Loading permissions...</p>;
                        if (error)
                            return <ErrorMessage error={error} />;
                        if (data)
                            return (
                                <Query query={ALL_USERS_QUERY} variables={{
                                    skip: this.props.page * perPage - perPage,
                                    first: perPage
                                }}>
                                    {payload => {
                                        if (payload.loading)
                                            return <p>Loading users...</p>;
                                        if (payload.data)
                                            return this.renderPermissionsGrid(data.permissions, payload.data.users);
                                    }}
                                </Query>
                            );
                    }}
                </Query>
            </NeedsPermission>
        );
    }
}

UserPermissions.propTypes = {
    user: PropTypes.shape({
        id: PropTypes.string,
        permissions: PropTypes.array,
    }),
    allPermissions: PropTypes.array,
};

Permissions.propTypes = {
    userData: PropTypes.number,
    page: PropTypes.number,
};

export default Permissions;
