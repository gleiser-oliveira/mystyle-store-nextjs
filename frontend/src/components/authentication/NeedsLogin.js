import React from 'react';
import PropTypes from 'prop-types';
import User from '../authentication/User';
import Signin from '../authentication/Signin';

const NeedsLogin = props => (
    <User>
        {payload => (
            payload.data.me ? props.children : <div>
                <p>You need to sign in to access this resource.</p>
                <Signin/>
            </div>
        )}
    </User>
);

NeedsLogin.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.object,
    ]),
};

export default NeedsLogin;
