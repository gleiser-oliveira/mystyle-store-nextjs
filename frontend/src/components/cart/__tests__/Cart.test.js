import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { InMemoryCache } from 'apollo-cache-inmemory';
import toJSON from 'enzyme-to-json';
import Cart from '../Cart';
import { LOCAL_STATE_QUERY } from '../../../lib/ClientGQL/LocalGQL';
import { CURRENT_USER_QUERY } from '../../authentication/AuthenticationOperationsGQL';
import { updateWrapper, fakeUser, fakeCart } from '../../../lib/testUtils';

// mocked local cache for LOCAL_STATE_QUERY
const cache = new InMemoryCache();
cache.writeData({
    data: {
        cartOpen: true,
    },
});

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: fakeCart(),
                },
            },
        },
    },
    {
        request: { query: LOCAL_STATE_QUERY },
        result: {
            data: {
                cartOpen: true,
            },
        },
    },
];

describe('testing the <Cart/> component', () => {
    it('renders', async() => {
        const wrapper = mount(
            <MockedProvider mocks={mocks} cache={cache} resolvers={{}}>
                <Cart userCart={ fakeCart() }/>
            </MockedProvider>
        );

        await updateWrapper(wrapper, 50);
        const cart = wrapper.find('div[open=true]');
        const header = cart.find('header');

        expect(toJSON(header)).toMatchSnapshot();
        expect(cart.find('CartProduct')).toHaveLength(2);
    });
});
