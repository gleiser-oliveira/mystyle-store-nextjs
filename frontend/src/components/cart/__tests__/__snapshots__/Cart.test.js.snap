// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the <Cart/> component renders 1`] = `
<header>
  <styles__CloseButton
    onClick={[Function]}
    title="close"
  >
    <button
      className="styles__CloseButton-sc-1tfdw1p-1 kgCIRv"
      onClick={[Function]}
      title="close"
    >
      Ã—
    </button>
  </styles__CloseButton>
  <styles__Supreme>
    <h3
      className="styles__Supreme-sc-1tfdw1p-2 eaAogB"
    >
      Your Cart
    </h3>
  </styles__Supreme>
  <p>
    You have 
    4
     products in your cart.
  </p>
</header>
`;
